Fix restriding logic for structured kernels

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #53973 Add some trivial meta functions for commonly used inplace operations
* #53972 Make storage access error NotImplementedError
* #53823 Compute type_equal() without reference to backend()
* #53816 Also propagate storage_access_should_throw_ when copying tensor metadata
* **#53759 Fix restriding logic for structured kernels**
* #53682 Add meta device to generic device testing framework, skip NotImplementedError
* #53747 Add 'noarch' tests which only run in one CI config

Fixes #53587, see issue for in-depth explanation of the bug.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D26971342](https://our.internmc.facebook.com/intern/diff/D26971342)

ezyang committed:
Fix restriding logic for structured kernels

Fixes #53587, see issue for in-depth explanation of the bug.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

bdhirsh commented on tools/codegen/dest/register_dispatch_key.py:
```diff
@@ -279,12 +280,17 @@ def gen_class_set_output_body(self, k: SchemaKind) -> str:
+// Only restride if a resize occurred; otherwise we ignore the (advisory)
+// strides from the meta function and directly use the output tensor's
+// preexisting strides
+if (resized) {{
```
So, the behavior now is basically "We now only use the input's memory format to determine strides when we have to resize the output", right? Posting a snippet below to confirm what I'm talking about. 
```
input = torch.ones((2, 2, 2, 2), memory_format=torch.channels_last)
output = torch.ones((2, 2, 2, 2))
# output doesn't get resized, so we don't clobber the output's strides/memory_format.
torch._C._nn.upsample_nearest2d(input, (2,2), out=output).is_contiguous(memory_format=torch.channels_last)
# returns False

output = torch.ones(0)
# output gets resized, so we DO restride according to the input's memory format
torch._C._nn.upsample_nearest2d(input, (2,2), out=output).is_contiguous(memory_format=torch.channels_last)
# returns True
```

That all seems reasonable, since:
- I can imagine if a user explicitly specifies an output with the correct size, they probably don't want it to get re-strided?
- We also don't allow you to pass in an output tensor with non-zero size if it **doesn't** match the expected output size. (Well, the behavior is deprecated anyway). So really your only options are to pass in a zero-sized tensor, indicating that you want whatever properties the input tensor has, or you specify your own output tensor with the right size.

Just calling it out explicitly, since I **think** the behavior of the `upsample` kernels before porting them to structured was that they unconditionally copied the suggested memory format of the input ([here](https://github.com/pytorch/pytorch/pull/52065/files#diff-10f54f9d1a5f77be88290a989c6e25da6c8d5aa931462c43f3a3cc8685d8d382L43)).

ngimel replied:
Expected out behavior is described here https://github.com/pytorch/pytorch/wiki/Developer-FAQ#how-does-out-work-in-pytorch, it is what you describe. Thanks for calling it out explicitly, and you are correct that previous behavior of `upsample` didn't conform to those rules.

ngimel commented on test/test_torch.py:
```diff
@@ -2591,23 +2591,53 @@ def test_empty_meta(self):
-            # TODO: this is not a sustainable way of testing meta functions,
-            # but I want some quick scaffolding first before a more
-            # integrated testing strategy
+            # TODO: this test should be triggered by test_nn.py but right
```
probably these tests belong in test_nn, but no need to change this PR.

ezyang committed:
Update on "Fix restriding logic for structured kernels"


Fixes #53587, see issue for in-depth explanation of the bug.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Fix restriding logic for structured kernels"


Fixes #53587, see issue for in-depth explanation of the bug.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D26971342](https://our.internmc.facebook.com/intern/diff/D26971342)

[ghstack-poisoned]

ezyang committed:
Update on "Fix restriding logic for structured kernels"


Fixes #53587, see issue for in-depth explanation of the bug.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D26971342](https://our.internmc.facebook.com/intern/diff/D26971342)

[ghstack-poisoned]

ezyang committed:
Update on "Fix restriding logic for structured kernels"


Fixes #53587, see issue for in-depth explanation of the bug.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D26971342](https://our.internmc.facebook.com/intern/diff/D26971342)

[ghstack-poisoned]

ezyang committed:
Update on "Fix restriding logic for structured kernels"


Fixes #53587, see issue for in-depth explanation of the bug.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D26971342](https://our.internmc.facebook.com/intern/diff/D26971342)

[ghstack-poisoned]

ezyang committed:
Update on "Fix restriding logic for structured kernels"


Fixes #53587, see issue for in-depth explanation of the bug.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D26971342](https://our.internmc.facebook.com/intern/diff/D26971342)

[ghstack-poisoned]
